storage/index.go

```diff
@@ -21,61 +21,260 @@
+)
+
+type dbIndexState byte
+
+const (
+	dbIndexStateOpen dbIndexState = iota
+	dbIndexStateClosed
 )
 
+type dbIndexInsertMode byte
+
+// nolint
+const (
+	dbIndexInsertSync dbIndexInsertMode = iota
+	dbIndexInsertAsync
+)
+
+type dbIndexBlock struct {
+	segment    mem.Segment
+	expiryTime time.Time
+}
+
 type dbIndex struct {
-	opts Options
+	sync.RWMutex
+	insertMode dbIndexInsertMode
+	state      dbIndexState
+	active     dbIndexBlock
+
+	insertQueue databaseIndexInsertQueue
+	metrics     dbIndexMetrics
+	opts        index.Options
 }
 
-func newDatabaseIndex(o Options) (databaseIndex, error) {
-	return &dbIndex{
-		opts: o,
-	}, nil
+// nolint: deadcode
+func newDatabaseIndex(
+	fn newDatabaseIndexInsertQueueFn,
+	opts index.Options,
+) (databaseIndex, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	seg, err := mem.New(1, opts.MemSegmentOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	idx := &dbIndex{
+		insertMode: dbIndexInsertAsync,
+		active: dbIndexBlock{
+			segment:    seg,
+			expiryTime: maxTime, // FOLLOWUP(prateek): undo hard-coding to infinite retention
+		},
+		metrics: newDatabaseIndexMetrics(opts.InstrumentOptions().MetricsScope()),
+		opts:    opts,
+	}
+
+	queue := fn(idx.writeBatch,
+		opts.ClockOptions().NowFn(),
+		opts.InstrumentOptions().MetricsScope())
+	if err := queue.Start(); err != nil {
+		return nil, err
+	}
+	idx.insertQueue = queue
+
+	return idx, nil
+}
+
+func (i *dbIndex) writeBatch(inserts []dbIndexInsert) error {
+	// NB(prateek): we use a read lock to guard against mutation of the
+	// dbIndexBlock currently active, mutations within the underlying
+	// dbIndexBlock are guarded by primitives internal to it.
+	i.RLock()
+	defer i.RUnlock()
+
+	if !i.isOpenWithRLock() {
+		// NB(prateek): skip calling any of the `OnIndexFinalize` methods
+		// on the provided inserts to terminate quicker during shutdown.
+		i.metrics.InsertAfterClose.Inc(int64(len(inserts)))
+		return errDbIndexUnableToWriteClosed
+	}
+
+	var numErr int64
+	for _, insert := range inserts {
+		// FOLLOWUP(prateek): need to query before insert to ensure no duplicates && add test
+		err := i.active.segment.Insert(insert.doc)
+		if err != nil {
+			numErr++
+		} else {
+			insert.fns.OnIndexSuccess(i.active.expiryTime)
+		}
+		// NB: we need to release held resources so we un-conditionally execute the Finalize.
+		insert.fns.OnIndexFinalize()
+	}
+
+	if numErr != 0 {
+		i.metrics.AsyncInsertErrors.Inc(numErr)
+	}
+
+	return nil
 }
 
 func (i *dbIndex) Write(
-	ctx context.Context,
 	namespace ident.ID,
 	id ident.ID,
-	tags ident.TagIterator,
+	tags ident.Tags,
+	fns indexInsertLifecycleHooks,
 ) error {
-	return errIndexingNotImplemented
+	d, err := i.doc(namespace, id, tags)
+	if err != nil {
+		fns.OnIndexFinalize()
+		return err
+	}
+
+	i.RLock()
+	if !i.isOpenWithRLock() {
+		i.RUnlock()
+		i.metrics.InsertAfterClose.Inc(1)
+		fns.OnIndexFinalize()
```

Comment:
Hm why do we finalize on error here but not in write batch method?  Also do we need a write batch and and non batched one?  Does this ever get used?

Reply:
At this point the write has not made it to the index queue. We've received the write and realised the index has been closed so we terminate early.

Reply:
Re: batch v non-batch. This thread explains the call chains: https://github.com/m3db/m3db/pull/514#discussion_r177185849 

Don't need batch v non-batch because all writes are batched internally by the queue.